clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3007,19 +3007,37 @@ class SYCLKernelNameTypeVisitor
         S.getASTContext().getLangOpts().SYCLUnnamedLambda;
     const DeclContext *DeclCtx = Tag->getDeclContext();
     if (DeclCtx && !UnnamedLambdaEnabled) {
-      auto *NameSpace = dyn_cast_or_null<NamespaceDecl>(DeclCtx);
-      if (NameSpace && NameSpace->isStdNamespace()) {
-        S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
-            << KernelNameType;
-        S.Diag(KernelInvocationFuncLoc, diag::note_invalid_type_in_sycl_kernel)
-            << /* kernel name cannot be a type in the std namespace */ 2
-            << QualType(Tag->getTypeForDecl(), 0);
-        IsInvalid = true;
-        return;
+      // Loop through Declaration Contexts to detect kernel names nested inside
+      // "std" and "anonymous" namespaces.
+      while (!DeclCtx->isTranslationUnit()) {
+        auto *NSDecl = dyn_cast<NamespaceDecl>(DeclCtx);
+        if (NSDecl && NSDecl->isStdNamespace()) {
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::err_sycl_kernel_incorrectly_named)
+              << KernelNameType;
+          S.Diag(KernelInvocationFuncLoc,
+                 diag::note_invalid_type_in_sycl_kernel)
+              << /* kernel name cannot be a type in the std namespace */ 2
+              << QualType(Tag->getTypeForDecl(), 0);
```

Comment:
This was also part of the original code, but why are we making a `QualType` here rather than passing `Tag` to the diagnostics engine and letting it format it as a named declaration?

Reply:
I didn't catch this, but yes, I can pass `Tag` to the diagnostics engine and it would print the invalid Tag type.
i.e. If '`std::NestedInStd::NestedStruct`' was the kernel name, passing `Tag`, would print, `NestedStruct` as the invalid type.

Reply:
Ah, I see! If you want to print a fully-qualified type, you can do that from the diagnostic definition itself with the `q` qualifier, as in: https://github.com/intel/llvm/blob/47b85e5603ae4966368b3c5137283b6fc8689809/clang/include/clang/Basic/DiagnosticParseKinds.td#L551

Reply:
Neat trick!  I didn't know that one!

Reply:
@AaronBallman 


> Ah, I see! If you want to print a fully-qualified type, you can do that from the diagnostic definition itself with the `q` qualifier, as in:
> 
> https://github.com/intel/llvm/blob/47b85e5603ae4966368b3c5137283b6fc8689809/clang/include/clang/Basic/DiagnosticParseKinds.td#L551

Passing '`Tag`' and using the `q `qualifier works well for most cases except for examples like below:

```
namespace std {
typedef struct {
} StructInStd;
} // namespace std
```

//kernel invocation call
```
q.submit([&](cl::sycl::handler &h) {
      h.single_task<std::StructInStd>([] {});
    });
```
  
//DiagnosticSemaKinds.td
`   def err_sycl_kernel_incorrectly_named : Error<"%0 is an invalid kernel name,type %q1 cannot be in the \'std\' namespace">;`
   
//SemaSYCL.cpp
```
   S.Diag(KernelInvocationFuncLoc, diag::err_sycl_kernel_incorrectly_named)
            << KernelNameType << Tag;
```
//Error    
`sycl.hpp Line 220: 'std::StructInStd' is an invalid kernel name,type 'std::(anonymous)' cannot be in the 'std' namespace`

Note that the type above gets printed as `std::(anonymous)` vs  `std::StructInStd` when  `QualType(Tag->getTypeForDecl(), 0);` is passed(without the `q `qualifier) instead of 'Tag'
This is because 'Tag' here is an empty string "" for this example.
I have update the code to use the `QualType(Tag->getTypeForDecl(), 0);`, to print the fully qualified type for all cases. Let me know if this is okay.

Reply:
I think there's some type vs declaration confusion here (possibly on my side). The code is checking where the type is declared, so to me, the warning is about the declaration of the type and not the type itself. Based on that, I think whatever we get when printing `Tag` is reasonable because that's the declaration and if that's confusing, we can use a note to note the location of the offending tag declaration.

This is one of those weird edge cases where the declaration has no name except that it does for linkage purposes. `TagDecl::getTypedefNameForAnonDecl()` could be used to print the typedef name

Reply:
> I think there's some type vs declaration confusion here

The original code had [getTypeForDecl()](https://clang.llvm.org/doxygen/Decl_8h_source.html#l03109) which returns the Type for that declaration, which led to me to think that we wanted the "Type" itself and not the "declaration" of the type.
I will switch back to using "Tag". Thanks.

Reply:
@erichkeane -- any preference from you? I presume use of the type before was perhaps accidental rather than purposeful, but you may have more context.

Reply:
I don't know any further context here.

Reply:
Okay, then I'd suggest sticking with the declaration rather than the type. If we find the diagnostic makes it hard to find the correct declaration in practice, we could attach another note that points to the declaration in question (but I'd say hold off on that until a user finds an issue).

Reply:
Does it matter to change "type" to "declaration/declaration of type" in the diagnostics text? Or leave "type" as is?

> sycl.hpp Line 220: 'std::StructInStd' is an invalid kernel name, **type** 'std::(anonymous)' cannot be in the 'std' namespace

Reply:
I think that's a great suggestion -- "declaration of type" make the diagnostic far more readable.

Reply:
Actually, I wonder if this is better still:

`'std::StructInStd' is an invalid kernel name; template argument 'std::(anonymous)' is declared in the 'std' namespace`

Reply:
I am okay with this, but I wonder if it will exceed the 80 column limit rule mentioned under best practices here ( especially if we have longer kernel names)
https://clang.llvm.org/docs/InternalsManual.html#the-format-string
```
Here are some “best practices” for writing the English format string:

Keep the string short. It should ideally fit in the 80 column limit of the DiagnosticKinds.td file. This avoids the diagnostic wrapping when printed, and forces you to think about the important point you are conveying with the diagnostic.
Take advantage of location information. The user will be able to see the line and location of the caret, so you don’t need to tell them that the problem is with the 4th argument to the function: just point to it.
Do not capitalize the diagnostic string, and do not end it with a period.
If you need to quote something in the diagnostic string, use single quotes.
```

Reply:
Hmm, I guess I don't think this is too long of a diagnostic, but you are correct that it's longer than the 80 col limit suggestion in the docs. If this is a worry, perhaps: `'std::StructInStd' is an invalid kernel name; 'std::(anonymous)' is declared in the 'std' namespace` (dropping the "template argument" from my earlier suggestion).

Reply:
I was following up on your earlier suggestion that "Diagnostic text should be succinct" and the original diagnostic text had ~103 character count (>80 col limit). I will go with this one - `'std::StructInStd' is an invalid kernel name; 'std::(anonymous)' is declared in the 'std' namespace`
I am not sure if telling users 'std::(anonymous)' **cannot be** declared in the 'std' namespace is more obvious/helpful or 'std::StructInStd' is an invalid kernel name; 'std::(anonymous)' **is declared** in the 'std' namespace is self explanatory

Reply:
> I was following up on your earlier suggestion that "Diagnostic text should be succinct" and the original diagnostic text had ~103 character count (>80 col limit). I will go with this one - 'std::StructInStd' is an invalid kernel name; 'std::(anonymous)' is declared in the 'std' namespace

SGTM!

> I am not sure if telling users 'std::(anonymous)' cannot be declared in the 'std' namespace is more obvious/helpful or 'std::StructInStd' is an invalid kernel name; 'std::(anonymous)' is declared in the 'std' namespace is self explanatory

FWIW, I think "cannot be declared" implies that the programmer did the declaration (which is highly unlikely to be the case with a declaration in `std`). "is declared" doesn't carry those connotations for me.

Reply:
I really hope users run into these errors often. I lost track of the number of times these underwent improvements ¯\_(ツ)_/¯